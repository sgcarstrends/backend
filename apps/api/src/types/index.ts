// Reexport all types from @sgcarstrends/types
export type {
  Car,
  COE,
  Collection,
  FuelType,
  LatestMonth,
  OrderBy,
  PQP,
  VehicleClass,
} from "@sgcarstrends/types";

// Local types that aren't shared
export type Make = string;

// Workflow types
export enum Stage {
  DEVELOPMENT = "dev",
  STAGING = "staging",
  PRODUCTION = "prod",
}
