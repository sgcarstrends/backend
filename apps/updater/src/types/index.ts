export type Stage = "dev" | "staging" | "prod";

export type {
  Car,
  COE,
  PQP,
} from "@sgcarstrends/types";
