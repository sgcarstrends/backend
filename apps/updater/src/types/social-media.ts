import type { LICreateResponse } from "linkedin-api-client";

export type CreatedEntityId = LICreateResponse["createdEntityId"];

interface PostToSocialMediaParam {
  message: string;
  link: string;
}

export interface PostToDiscordParam extends PostToSocialMediaParam {}
export interface PostToLinkedInParam extends PostToSocialMediaParam {}
export interface PostToTelegramParam extends PostToSocialMediaParam {}
export interface PostToTwitterParam extends PostToSocialMediaParam {}

export type ResharePostParam = {
  createdEntityId: CreatedEntityId;
  message: string;
};
