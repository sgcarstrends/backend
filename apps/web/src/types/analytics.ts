export interface AnalyticsData {
  totalViews: number;
  uniqueVisitors: number;
  topCountries: Array<{
    country: string;
    flag: string;
    count: number;
  }>;
  topCities: Array<{
    city: string;
    country: string;
    flag: string;
    count: number;
  }>;
  topPages: Array<{
    pathname: string;
    count: number;
  }>;
  topReferrers: Array<{
    referrer: string;
    count: number;
  }>;
  dailyViews: Array<{
    date: string;
    count: number;
  }>;
}