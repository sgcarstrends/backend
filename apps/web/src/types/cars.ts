interface Metric {
  label: string;
  count: number;
}

interface Data {
  period: string;
  total: number;
  fuelType: Metric[];
  vehicleType: Metric[];
}

export interface RegistrationStat {
  name: string;
  count: number;
}

export interface Registration {
  month: string;
  total: number;
  fuelType: RegistrationStat[];
  vehicleType: RegistrationStat[];
}

export interface Comparison {
  currentMonth: Data;
  previousMonth: Data;
  previousYear: Data;
}

export interface TopType {
  month: string;
  topFuelType: {
    name: string;
    total: number;
  };
  topVehicleType: {
    name: string;
    total: number;
  };
}

export interface Make {
  make: string;
  count: number;
}

export interface FuelType {
  fuelType: string;
  total: number;
  makes: Make[];
}

export interface TopMake {
  month: string;
  topFuelType: {
    name: string;
    total: number;
  };
  topVehicleType: {
    name: string;
    total: number;
  };
}
