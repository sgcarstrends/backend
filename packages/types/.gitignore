# Build outputs
dist/

# Testing
coverage/

# Cache
*.tsbuildinfo