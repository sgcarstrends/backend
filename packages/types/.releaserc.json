{
  "extends": "semantic-release-monorepo",
  "branches": ["main"],
  "tagFormat": "@sgcarstrends/types@${version}",
  "plugins": [
    "@semantic-release/commit-analyzer",
    "@semantic-release/release-notes-generator",
    [
      "@semantic-release/changelog",
      {
        "changelogFile": "CHANGELOG.md"
      }
    ],
    [
      "@semantic-release/npm",
      {
        "npmPublish": false
      }
    ],
    [
      "@semantic-release/git",
      {
        "assets": ["CHANGELOG.md", "package.json"],
        "message": "chore(types): release ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}"
      }
    ],
    [
      "@semantic-release/github",
      {
        "successComment": "ðŸŽ‰ This change has been released in `@sgcarstrends/types@${nextRelease.version}`"
      }
    ]
  ]
}
