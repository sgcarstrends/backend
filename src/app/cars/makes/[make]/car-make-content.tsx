import Image from "next/image";
import { columns } from "@/app/cars/makes/[make]/columns";
import { TrendChart } from "@/app/cars/makes/[make]/trend-chart";
import { LastUpdated } from "@/components/last-updated";
import { MakeSelector } from "@/components/make-selector";
import NoData from "@/components/no-data";
import Typography from "@/components/typography";
import {
  Card,
  CardContent,
  CardDescription,
  CardHeader,
  CardTitle,
} from "@/components/ui/card";
import { DataTable } from "@/components/ui/data-table";
import { UnreleasedFeature } from "@/components/unreleased-feature";
import { type Car, type Make } from "@/types";

interface CarMakeContentProps {
  make: string;
  cars: { make: string; total: number; data: Car[] };
  makes: Make[];
  lastUpdated?: number | null;
  logoUrl?: string | null;
}

export function CarMakeContent({
  make,
  cars,
  makes,
  lastUpdated,
  logoUrl,
}: CarMakeContentProps) {
  if (!cars) {
    return <NoData />;
  }

  return (
    <div className="flex flex-col gap-4">
      <div className="flex flex-col gap-2">
        <div className="flex flex-col justify-between gap-2 lg:flex-row lg:items-center">
          <div className="flex flex-col items-center">
            <UnreleasedFeature>
              {logoUrl && (
                <Image
                  alt={`${cars.make} logo`}
                  src={logoUrl}
                  width={128}
                  height={128}
                />
              )}
            </UnreleasedFeature>
            <Typography.H1>{cars.make}</Typography.H1>
          </div>
          <div className="flex flex-row-reverse items-center justify-between gap-2 lg:flex-row">
            {lastUpdated && <LastUpdated lastUpdated={lastUpdated} />}
            <MakeSelector makes={makes} selectedMake={make} />
          </div>
        </div>
      </div>

      <Card>
        <CardHeader>
          <CardTitle>Historical Trend</CardTitle>
          <CardDescription>Past registrations</CardDescription>
        </CardHeader>
        <CardContent>
          <TrendChart data={cars.data.toReversed()} />
        </CardContent>
      </Card>
      <Card>
        <CardHeader>
          <CardTitle>Summary</CardTitle>
          <CardDescription>
            Breakdown of fuel &amp; vehicle types by month
          </CardDescription>
        </CardHeader>
        <CardContent>
          <DataTable columns={columns} data={cars.data} />
        </CardContent>
      </Card>
    </div>
  );
}
