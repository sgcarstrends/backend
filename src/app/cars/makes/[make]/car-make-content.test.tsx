import type { ReactNode } from "react";
import { render, screen } from "@testing-library/react";
import { type Car, type Make } from "@/types";
import { CarMakeContent } from "./car-make-content";

vi.mock("next/navigation", () => ({
  useRouter: () => ({
    push: vi.fn(),
    back: vi.fn(),
  }),
}));

vi.mock("@/app/cars/makes/[make]/trend-chart", () => ({
  TrendChart: () => <div>TrendChart</div>,
}));

vi.mock("@/components/make-selector", () => ({
  MakeSelector: () => <div>MakeSelector</div>,
}));

vi.mock("@/components/ui/data-table", () => ({
  DataTable: () => <div>DataTable</div>,
}));

vi.mock("@/components/unreleased-feature", () => ({
  UnreleasedFeature: ({ children }: { children: ReactNode }) => (
    <div>{children}</div>
  ),
}));

const mockCars = {
  make: "TOYOTA",
  total: 100,
  data: [
    {
      month: "2024-01",
      make: "TOYOTA",
      fuelType: "PETROL",
      vehicleType: "SALOON",
      number: 50,
    },
    {
      month: "2024-02",
      make: "TOYOTA",
      fuelType: "HYBRID",
      vehicleType: "SUV",
      number: 30,
    },
  ] as unknown as Car[],
};

const mockMakes: Make[] = ["TOYOTA", "HONDA", "BMW"];

const mockLastUpdated = 1735660800; // 2025-01-01 00:00:00 UTC

describe("CarMakeContent", () => {
  it("renders make name", () => {
    render(<CarMakeContent make="TOYOTA" cars={mockCars} makes={mockMakes} />);
    expect(screen.getByText("TOYOTA")).toBeVisible();
  });

  it("renders Historical Trend card", () => {
    render(<CarMakeContent make="TOYOTA" cars={mockCars} makes={mockMakes} />);
    expect(screen.getByText("Historical Trend")).toBeVisible();
    expect(screen.getByText("Past registrations")).toBeVisible();
  });

  it("renders Summary card", () => {
    render(<CarMakeContent make="TOYOTA" cars={mockCars} makes={mockMakes} />);
    expect(screen.getByText("Summary")).toBeVisible();
    expect(
      screen.getByText("Breakdown of fuel & vehicle types by month"),
    ).toBeVisible();
  });

  it("renders NoData component when cars is null", () => {
    render(
      <CarMakeContent make="TOYOTA" cars={null as any} makes={mockMakes} />,
    );
    expect(screen.getByText("No Data Available")).toBeVisible();
  });

  it("renders LastUpdated component when lastUpdated is provided", () => {
    render(
      <CarMakeContent
        make="TOYOTA"
        cars={mockCars}
        makes={mockMakes}
        lastUpdated={mockLastUpdated}
      />,
    );
    expect(screen.getByText(/Last updated:/)).toBeVisible();
  });

  it("renders make logo image", () => {
    render(<CarMakeContent make="TOYOTA" cars={mockCars} makes={mockMakes} />);
    const image = screen.getByRole("img");
    expect(image).toHaveAttribute("alt", "Logo");
    expect(image).toHaveAttribute("src", expect.stringContaining("TOYOTA.png"));
  });
});
