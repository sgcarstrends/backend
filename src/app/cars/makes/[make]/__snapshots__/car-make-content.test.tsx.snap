// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CarMakeContent > should render correctly 1`] = `
<div>
  <div
    class="flex flex-col gap-4"
  >
    <div
      class="flex flex-col gap-2"
    >
      <div
        class="flex flex-col justify-between gap-2 lg:flex-row lg:items-center"
      >
        <div
          class="flex flex-col items-center"
        >
          <div>
            <img
              alt="Logo"
              data-nimg="1"
              decoding="async"
              height="128"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Fassets.sgcarstrends.com%2Flogos%2FTOYOTA.png&w=256&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fassets.sgcarstrends.com%2Flogos%2FTOYOTA.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fassets.sgcarstrends.com%2Flogos%2FTOYOTA.png&w=256&q=75 2x"
              style="color: transparent;"
              width="128"
            />
          </div>
          <h1
            class="scroll-m-20 text-4xl font-extrabold tracking-tight lg:text-5xl"
          >
            TOYOTA
          </h1>
        </div>
        <div
          class="flex flex-row-reverse items-center justify-between gap-2 lg:flex-row"
        >
          <div
            class="text-muted-foreground text-xs"
          >
            Last updated:
             
            <span
              class="underline"
            >
              21 Jan 1970, 9:37AM
            </span>
          </div>
          <div>
            MakeSelector
          </div>
        </div>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-[data-slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
        data-slot="card-header"
      >
        <div
          class="leading-none font-semibold"
          data-slot="card-title"
        >
          Historical Trend
        </div>
        <div
          class="text-muted-foreground text-sm"
          data-slot="card-description"
        >
          Past registrations
        </div>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div>
          TrendChart
        </div>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-[data-slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6"
        data-slot="card-header"
      >
        <div
          class="leading-none font-semibold"
          data-slot="card-title"
        >
          Summary
        </div>
        <div
          class="text-muted-foreground text-sm"
          data-slot="card-description"
        >
          Breakdown of fuel & vehicle types by month
        </div>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div>
          DataTable
        </div>
      </div>
    </div>
  </div>
</div>
`;
